### PR TITLE
Fix(formatter): preserve generics trailing comma in arrow function of tsx

### PR DESCRIPTION
**Pull request by cola119**
## Summary

This PR fixes https://github.com/rome/tools/issues/1295

As you probably know, TSX arrow functions must have a trailing comma of generics when it has a single type annotation, which means `TSTypeParameterDeclaration` need to know its grandparent node but the current node structure can't retrieve ancestors as far as I know. Eventually, I tried to solve this by changing `JSFunctionHead` (the middle node). Correct me if I wrong or you have any better solution.

This patch is based on the prettier implementation (https://github.com/prettier/prettier/issues/6114).

## Test Plan

This adds some tests of both tsx and typescript.

### CODE REVIEW

**Comment by ematipico** [comment]
I think, here in Rome, we can apply this fix at parser level. I *think* in prettier they had to do it that way because they don't control the parser. Here we can. 

The parser has access to the `path` which contains all the information about the file that is processing. I believe it would be a better fix. What do you think?

**Comment by cola119** [comment]
@ematipico Thank you for your review. That makes sense so I'll try to modify the parser per se.

File: internal/formatter/test-fixtures/tsx/generics/input.test.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+type Type1<T> = "type1";
+
+type Type2 = Two<T>;
+
+```
+
+### `Output`
+
+```tsx
+const f1 = <T,>(arg: T) => false;
```
Comment (cola119) [inline internal/formatter/test-fixtures/tsx/generics/input.test.md:44]: This is the main purpose.

**Comment by cola119** [comment]
@ematipico I fixed `js-parser`. Could you review it again?

File: internal/formatter/builders/js/typescript/TSTypeParameterDeclaration.ts [hunk 2]
```diff
@@ -33,7 +33,13 @@ export default function TSTypeParameterDeclaration(
 		params[0].type !== "TSMappedType";
 
 	if (shouldInline) {
-		return concat(["<", builder.tokenize(params[0], node), ">"]);
+		const hasTrailingComma = "hasTrailingComma" in node && node.hasTrailingComma;
```
Comment (ematipico) [inline internal/formatter/builders/js/typescript/TSTypeParameterDeclaration.ts:36]: Wasn't `!!node.hasTrailingComma` sufficient? Adding a string to check a boolean doesn't seem ideal

**Review by ematipico — APPROVED** [review]
A better solution from my point of view! Thank you very much! Just left a comment

File: internal/formatter/builders/js/typescript/TSTypeParameterDeclaration.ts (hunk 2 above)
Comment (cola119) [reply internal/formatter/builders/js/typescript/TSTypeParameterDeclaration.ts:36]: `node` is defined as `TSTypeParameterDeclaration | TSTypeParameterInstantiation` and `TSTypeParameterInstantiation` doesn't have `hasTrailingComma` props. 
image

So we need to narrow with in operator, I think.
https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-the-in-operator

**Pull request merged**